### ISSUE TITLE
Errors with custom acquisition function in autograd computation (inside gen.py)

### ISSUE THREAD

**Issue by natarajanmolecule** [pytorch/botorch#520]
I have been working on a Bayesian optimization problem for a chemical application. I am working with two objectives coming from two GP models. These models have different number of input dimensions, different number of training data points etc. Based on the manuals, I wrote my own qEHVI function by considering these two models as a model list, and overloading the forward method to compute posteriors from each of the models correctly. I stack up the samples from the sampler to pass into the next steps.. Just the acquisition functions works well in testing.

When I try it out with optimizer (optimize_acqf), I am running into an error saying that the "One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior." Is it happening due to the reference to the models present in the acquisition function even though I have my own forward function for the posteriors. Is the autograd function limiting this custom acquisition function? Any thoughts will be helpful.

## System Info
Please provide information about your setup, including
- BoTorch Version 0.3.0
- GPyTorch Version 1.1.1
- PyTorch Version 1.5.0
- Computer OS Wndows 10

**Comment by Balandat** [comment]
Sounds like an interesting use case. Could you provide a minimal example on toy data that reproduced this issue? Otherwise it's hard to say what's going on here.

**Comment by natarajanmolecule** [comment]
Thanks for the prompt reply Max. I will put a toy example together and send it along sometime later today.

**Comment by natarajanmolecule** [comment]
Here is a toy problem with two models, one containing 2 dimensions (20 training data pts) and another with 4 dimensions (30 training data pts), and I am performing BO with respect to 2 dimensions (for now other dimensions are zero). I am getting the same error in this case too. Any help is greatly appreciated.


```
# Updated qEIHypervolume to handle different number of inputs in the models
# Also import different GP models to create a list of models to be
# used for Bayesian Optimization

import gpytorch
import makeGP
import torch
from botorch.models.model_list_gp_regression import ModelListGP

from botorch.fit import fit_gpytorch_model
from botorch.models import SingleTaskGP
from botorch.utils import standardize
from gpytorch.mlls import ExactMarginalLogLikelihood


# generate synthetic data
X1 = torch.rand(20, 2)
Y1 = torch.sin(X1[:, 0]) + torch.cos(X1[:, 1])
X2 = torch.rand(30,4)
Y2 = torch.sin(X2[:, 0])**2 + torch.cos(X2[:, 1]) * torch.sin(X2[:,2]) + torch.tan(X2[:,3])
Y1 = standardize(Y1.unsqueeze(1))  # standardize to zero mean unit variance
Y2 = standardize(Y2.unsqueeze(1))

# construct and fit the multi-output model
gp1 = SingleTaskGP(X1, Y1)
mll1 = ExactMarginalLogLikelihood(gp1.likelihood, gp1)
fit_gpytorch_model(mll1)

gp2 = SingleTaskGP(X2, Y2)
mll2 = ExactMarginalLogLikelihood(gp2.likelihood, gp2)
fit_gpytorch_model(mll2)

tkwargs = {
    "dtype": torch.double,
    "device": torch.device("cuda" if torch.cuda.is_available() else "cpu"),
}

models_gp = []
inputNames = ['X1','X2']

models_gp = [gp1,gp2]

modelList = ModelListGP(*models_gp)

from typing import Callable, List, Optional
import pandas as pd
import torch
from botorch.acquisition.acquisition import AcquisitionFunction
from botorch.acquisition.multi_objective.objective import (
    IdentityMCMultiOutputObjective,
    MCMultiOutputObjective,
)
from botorch.exceptions.errors import UnsupportedError
from botorch.models.model import Model
from botorch.sampling.samplers import MCSampler, SobolQMCNormalSampler
from botorch.utils.multi_objective.box_decomposition import NondominatedPartitioning
from botorch.utils.objective import apply_constraints_nonnegative_soft
from botorch.utils.torch import BufferDict
from botorch.utils.transforms import concatenate_pending_points, t_batch_mode_transform
from botorch.acquisition.multi_objective.monte_carlo import qExpectedHypervolumeImprovement
from botorch.utils.transforms import concatenate_pending_points, t_batch_mode_transform
from torch import Tensor
from botorch.posteriors.gpytorch import GPyTorchPosterior
from gpytorch.distributions import MultitaskMultivariateNormal, MultivariateNormal
from botorch.models.utils import (
    _make_X_full,
    add_output_dim,
    gpt_posterior_settings,
    mod_batch_shape,
    multioutput_to_batch_mode_transform,
)

class qExpectedHypervolumeImprovement_new(qExpectedHypervolumeImprovement):

    def __init__(self, model: Model,
        ref_point: List[float],
        partitioning: NondominatedPartitioning,
        sampler: Optional[MCSampler] = None,
        objective: Optional[MCMultiOutputObjective] = None,
        constraints: Optional[List[Callable[[Tensor], Tensor]]] = None,
        X_pending: Optional[Tensor] = None,
        eta: float = 1e-3,
        neededInputs=None,
        modelParamsList=None) -> None:
        super().__init__(
            model=model,
            ref_point=ref_point,
            partitioning=partitioning,
            sampler=sampler,
            objective=objective,
            constraints=constraints,
            X_pending=X_pending,
            eta=eta,
        )
        self.neededInputs = neededInputs
        self.modelParamsList = modelParamsList #List of list of model parameters

    @property
    def neededInputs(self):
        return self._neededInputs
    @neededInputs.setter
    def neededInputs(self,value):
        self._neededInputs = value
    @property
    def modelParamsList(self):
        return self._modelParamsList
    @modelParamsList.setter
    def modelParamsList(self,value):
        self._modelParamsList = value
    
    # Updated forward computation for our usecase
    @concatenate_pending_points
    @t_batch_mode_transform()
    def forward(self, X: Tensor) -> Tensor:

        # First create n different model posteriors
        nmodels = len(self.model.models)
        mvns = []
        sampleslist = []
        for model in range(nmodels):
            inputNames = self.neededInputs
            # inputValues = X.detach().numpy()[:,:,0]
            inputValues = np.squeeze(X.detach().numpy())
            inputDF = pd.DataFrame(data=inputValues,columns=inputNames)
            XNames = self.modelParamsList[model]
            testDF = pd.DataFrame(columns=XNames)

            index_dict = dict((value, idx) for idx, value in enumerate(XNames))
            input_index_dict = dict((value, idx) for idx, value in enumerate(inputNames))
            inter = set(index_dict).intersection(inputNames)
            inputIndices = [index_dict[x] for x in inter]
            input_index_dict = [input_index_dict[x] for x in inter]
            testDF[testDF.keys()[inputIndices]] = inputDF.iloc[:,input_index_dict]
            # testDF[testDF.keys()[0:-3]] = testDF[testDF.keys()[0:-3]] / 100.0
            testDF.fillna(0, inplace=True)
            inputForPost = testDF.values
            inputForPostTensor = torch.from_numpy(inputForPost).type(torch.FloatTensor)

            self.model.models[model].eval()
            with gpt_posterior_settings():
                mvn = self.model.models[model](inputForPostTensor)
                mvn = self.model.models[model].likelihood(mvn, inputForPostTensor)
            mvns.append(mvn)
            posterior = GPyTorchPosterior(mvn=mvn)

            samples = self.sampler(posterior)
            sampleslist.append(samples)

        # posterior = GPyTorchPosterior(mvn=MultitaskMultivariateNormal.from_independent_mvns(mvns=mvns))
        # posterior = self.model.posterior(X)
        # samples = self.sampler(posterior)
        allsamples = torch.stack(sampleslist,dim=3)
        return self._compute_qehvi(samples=allsamples)

from botorch.acquisition.monte_carlo import qExpectedImprovement, qNoisyExpectedImprovement
from botorch.sampling.samplers import SobolQMCNormalSampler
from botorch.exceptions import BadInitialCandidatesWarning
from botorch.utils.multi_objective.pareto import is_non_dominated
from botorch.utils.multi_objective.hypervolume import Hypervolume

MC_SAMPLES = 30

qehvi_sampler = SobolQMCNormalSampler(num_samples=MC_SAMPLES)

import numpy as np

# Use original train data to identify the initial pareto
train_obj = torch.stack([Y1, Y2[0:len(Y1),:]], -1)


from botorch.utils.multi_objective.box_decomposition import NondominatedPartitioning
from botorch.acquisition.multi_objective.monte_carlo import qExpectedHypervolumeImprovement


partitioning = NondominatedPartitioning(num_outcomes=2, Y=train_obj)

acq_func = qExpectedHypervolumeImprovement_new(
        model=modelList,
        ref_point=[0.0,0.0],  # use known reference point 
        partitioning=partitioning,
        sampler=qehvi_sampler,
        neededInputs=inputNames,
        modelParamsList=[['X1','X2'],['X1','X2','X3','X4']]
    )

from botorch.optim.optimize import optimize_acqf, optimize_acqf_list
from botorch.acquisition.objective import GenericMCObjective
from botorch.utils.multi_objective.scalarization import get_chebyshev_scalarization
from botorch.utils.multi_objective.box_decomposition import NondominatedPartitioning
from botorch.acquisition.multi_objective.monte_carlo import qExpectedHypervolumeImprovement
from botorch.utils.sampling import sample_simplex


BATCH_SIZE = 4
standard_bounds = torch.zeros(2, len(inputNames), **tkwargs)
standard_bounds[1] = 1

candidates, _ = optimize_acqf(
        acq_function=acq_func,
        bounds=standard_bounds,
        q=BATCH_SIZE,
        num_restarts=20,
        raw_samples=1024,  # used for intialization heuristic
        options={"batch_limit": 5, "maxiter": 200, "nonnegative": True},
        sequential=True,
    )
print(candidates)
```

**Comment by Balandat** [comment]
Thanks for the repro. It looks like you're breaking up the compute graph by doing 
```python
inputValues = np.squeeze(X.detach().numpy())
inputDF = pd.DataFrame(data=inputValues,columns=inputNames)`
```

In order to be able to auto-differentiate through the acquisition function (see [this post](https://blog.paperspace.com/pytorch-101-understanding-graphs-and-automatic-differentiation/) for a discussion), you will need to only use torch operations, so the compute graph can be built. `X.detach()` explicitly detaches your computations from the existing graph built under the hood so far, which is why you're seeing the error that you're getting. You will have to swap out any non-torch operations in your forward pass (`np`, `pandas`, ...) with pytorch operators. 

Closing this issue for now, let us know if you still have difficulties getting this to work.
